From patchwork Fri Apr 21 14:38:39 2023
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: =?utf-8?q?Jan_Ekstr=C3=B6m?= <jeebjp@gmail.com>
X-Patchwork-Id: 41293
Delivered-To: ffmpegpatchwork2@gmail.com
Received: by 2002:a05:6a20:4645:b0:e3:3194:9d20 with SMTP id eb5csp1366280pzb;
        Fri, 21 Apr 2023 07:38:58 -0700 (PDT)
X-Google-Smtp-Source: 
 AKy350ZgqrsHgaKsC+BA1yms+NdB2VkPLzSTKKxPR7IbQUh0+TiP/WabdrCCy/mnZCl3UF1tOqEm
X-Received: by 2002:aa7:cb59:0:b0:504:aeb5:89c3 with SMTP id
 w25-20020aa7cb59000000b00504aeb589c3mr4762843edt.5.1682087937920;
        Fri, 21 Apr 2023 07:38:57 -0700 (PDT)
ARC-Seal: i=1; a=rsa-sha256; t=1682087937; cv=none;
        d=google.com; s=arc-20160816;
        b=P7h38DsFxFsDIF2Z7eKMvsACwQgDzVaRAyoV8RnI0aZDuDytkXrEOZSl6uqwDCRgnT
         HrA958bc5p4ZRXqzLmDMFlcPspRk00jBjSAS+A6vyd0p4fzi3aeqiS5AVwgZRsUiRUcK
         XpKXZijwHig99uY8FHv0lGehF4msBcXzfgq6qpz9JNpJ6AF+DijCslRSw2wyCpUJkZnI
         8rJGvYkVxKni9RgnNrQFzVErodsBHxtIWvF6wh0nxSiyCbiL3klVMQ786ZVan+Go8hOk
         Mz7aq3jTkm5MfWUsy6fgOkU9phQqEZ2qpDbgiPKWrN+MYtat9lPgAMB8zq/zF32XO6HA
         4RGw==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=google.com;
 s=arc-20160816;
        h=sender:errors-to:content-transfer-encoding:reply-to:list-subscribe
         :list-help:list-post:list-archive:list-unsubscribe:list-id
         :precedence:subject:mime-version:message-id:date:to:from
         :dkim-signature:delivered-to;
        bh=NYw9GuLaoAZeCa6lbro6vUcRsqqPu+d/VBLfIp9rRQo=;
        b=NClPXi9Jj8WMY2L4vYIbsTxvprUSYWv09J1TwAHZs+WYHzXKNWTRkkkPaqlkjpdjgO
         UPf9fbjO5RT03It2uBuKoL8ldTZJK6S1yhg3DDXdbB0s5kTh0BtU1Alcmz4ae233/Fle
         YizcdunpFKThfH1UWo5DHDXL+BmvD+9iviCsUWqxTCPiM5/o0n8juysVm5I4/lVhBsOE
         WJgC+KB65wFkP5sZs0Q4Qarxj2y9rvvRoRe4/X62+trwWjfkYPcuds1ARuPz9mVdZARZ
         7jb9w3tErjVYmVw2rIqZPnVqgtKnzFqQkMPBSSXpdzLuffEL0Nrh0yxB3yykHGGJE/AZ
         62Bw==
ARC-Authentication-Results: i=1; mx.google.com;
       dkim=neutral (body hash did not verify) header.i=@gmail.com
 header.s=20221208 header.b=jpzjMnml;
       spf=pass (google.com: domain of ffmpeg-devel-bounces@ffmpeg.org
 designates 79.124.17.100 as permitted sender)
 smtp.mailfrom=ffmpeg-devel-bounces@ffmpeg.org;
       dmarc=fail (p=NONE sp=QUARANTINE dis=NONE) header.from=gmail.com
Return-Path: <ffmpeg-devel-bounces@ffmpeg.org>
Received: from ffbox0-bg.mplayerhq.hu (ffbox0-bg.ffmpeg.org. [79.124.17.100])
        by mx.google.com with ESMTP id
 j16-20020a50ed10000000b00506b1acb507si3391828eds.464.2023.04.21.07.38.57;
        Fri, 21 Apr 2023 07:38:57 -0700 (PDT)
Received-SPF: pass (google.com: domain of ffmpeg-devel-bounces@ffmpeg.org
 designates 79.124.17.100 as permitted sender) client-ip=79.124.17.100;
Authentication-Results: mx.google.com;
       dkim=neutral (body hash did not verify) header.i=@gmail.com
 header.s=20221208 header.b=jpzjMnml;
       spf=pass (google.com: domain of ffmpeg-devel-bounces@ffmpeg.org
 designates 79.124.17.100 as permitted sender)
 smtp.mailfrom=ffmpeg-devel-bounces@ffmpeg.org;
       dmarc=fail (p=NONE sp=QUARANTINE dis=NONE) header.from=gmail.com
Received: from [127.0.1.1] (localhost [127.0.0.1])
	by ffbox0-bg.mplayerhq.hu (Postfix) with ESMTP id 653A868BF8A;
	Fri, 21 Apr 2023 17:38:53 +0300 (EEST)
X-Original-To: ffmpeg-devel@ffmpeg.org
Delivered-To: ffmpeg-devel@ffmpeg.org
Received: from mail-lf1-f47.google.com (mail-lf1-f47.google.com
 [209.85.167.47])
 by ffbox0-bg.mplayerhq.hu (Postfix) with ESMTPS id 91071680068
 for <ffmpeg-devel@ffmpeg.org>; Fri, 21 Apr 2023 17:38:46 +0300 (EEST)
Received: by mail-lf1-f47.google.com with SMTP id
 2adb3069b0e04-4ec94eb6dcaso1813821e87.3
 for <ffmpeg-devel@ffmpeg.org>; Fri, 21 Apr 2023 07:38:46 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=gmail.com; s=20221208; t=1682087925; x=1684679925;
 h=content-transfer-encoding:mime-version:message-id:date:subject:to
 :from:from:to:cc:subject:date:message-id:reply-to;
 bh=/gdW3zxVMj9LywJPWtxdlbwNuxZPeKsvOzQeZb39RkU=;
 b=jpzjMnmleDABmSPOCrQxpW/DJvJ3ATbNUBWBShxY3oz6KsI9nFW3C9qWvgEPZJn54K
 0JXCgNDsJKeX5OsKJco5giB/wPTOaKk1P8yua2N1Dhaj1wq391sTrZnr8KjbwJrokU1w
 9hWb1gxDJR5VsAIZWcWugeZ8OF6fWSg1dlEMQ+J6NJkJ69SOGmWSwEbI3zLlqHjTjEDy
 wT4tfyJKG2Oab3S+QyuzwrGQSJBtBf4gjyVNXzSPamxm/chgGc9ydLZg4TCfoLcWikfq
 7TsK8yUKd8lv58Yeauw/cxdErWg3K3++JrDknMyyGyohuawVkN87MwpHWFbyxvE6vAKz
 rf0A==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=1e100.net; s=20221208; t=1682087925; x=1684679925;
 h=content-transfer-encoding:mime-version:message-id:date:subject:to
 :from:x-gm-message-state:from:to:cc:subject:date:message-id:reply-to;
 bh=/gdW3zxVMj9LywJPWtxdlbwNuxZPeKsvOzQeZb39RkU=;
 b=DsFng3Nmuf3fmx/RY3whUSw+ezVnsOYsIcXESvc6nGOQibuJxXETAjsvMt3sG4iiRU
 FeOh/ePvbWRh3cDb1xF+pTbfNkhSmEUXygCl93SyaM8BDbvXw7JdhdSfc9p4ELWBSLug
 J/fFBmxrhUtLkcIDVM+qiIl/AUNdDuxr1+BNQkR/4LVeQIh1S9UVem2J6h52rRuxHTHy
 IVKgaFsedLRuRWfEakc+at7mDcXhthE/VqKtER65wl8t25BIxYi/Hqij8gyPMhFACR5+
 TYyI7/oS4oPCg13Nh0pYbIgCsr5KV43VFbotzfYl6OHZ66ZTONj2oZonKnZTe+s/Tr3q
 Rf7Q==
X-Gm-Message-State: AAQBX9dOs7+T+OcGI6bcoT2C3KnRR92bDfBEJ/IapA7ooFPijM7gRn4w
 hzYpS+EcI1IdWzv7esDSafnqe1IWIuQ=
X-Received: by 2002:ac2:44af:0:b0:4ee:d4bd:3475 with SMTP id
 c15-20020ac244af000000b004eed4bd3475mr1395834lfm.32.1682087925298;
 Fri, 21 Apr 2023 07:38:45 -0700 (PDT)
Received: from localhost.localdomain (91-153-198-187.elisa-laajakaista.fi.
 [91.153.198.187]) by smtp.gmail.com with ESMTPSA id
 j24-20020ac24558000000b004db3eff4b12sm577654lfm.171.2023.04.21.07.38.44
 for <ffmpeg-devel@ffmpeg.org>
 (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
 Fri, 21 Apr 2023 07:38:44 -0700 (PDT)
From: =?utf-8?q?Jan_Ekstr=C3=B6m?= <jeebjp@gmail.com>
To: ffmpeg-devel@ffmpeg.org
Date: Fri, 21 Apr 2023 17:38:39 +0300
Message-Id: <20230421143839.11839-1-jeebjp@gmail.com>
X-Mailer: git-send-email 2.40.0
MIME-Version: 1.0
Subject: [FFmpeg-devel] [PATCH] avformat/mov: enable identifying TTML
 subtitle streams as such
X-BeenThere: ffmpeg-devel@ffmpeg.org
X-Mailman-Version: 2.1.29
Precedence: list
List-Id: FFmpeg development discussions and patches <ffmpeg-devel.ffmpeg.org>
List-Unsubscribe: <https://ffmpeg.org/mailman/options/ffmpeg-devel>,
 <mailto:ffmpeg-devel-request@ffmpeg.org?subject=unsubscribe>
List-Archive: <https://ffmpeg.org/pipermail/ffmpeg-devel>
List-Post: <mailto:ffmpeg-devel@ffmpeg.org>
List-Help: <mailto:ffmpeg-devel-request@ffmpeg.org?subject=help>
List-Subscribe: <https://ffmpeg.org/mailman/listinfo/ffmpeg-devel>,
 <mailto:ffmpeg-devel-request@ffmpeg.org?subject=subscribe>
Reply-To: FFmpeg development discussions and patches <ffmpeg-devel@ffmpeg.org>
Errors-To: ffmpeg-devel-bounces@ffmpeg.org
Sender: "ffmpeg-devel" <ffmpeg-devel-bounces@ffmpeg.org>
X-TUID: EmnMi5zKl5BE

From: Jan Ekström <jan.ekstrom@24i.com>

The contents are full TTML XML documents. TTML writing tests'
results are updated as the streams are now properly identified
as TTML ones.

Signed-off-by: Jan Ekström <jan.ekstrom@24i.com>
---
 libavformat/mov.c                | 11 +++++++++--
 tests/ref/fate/mov-mp4-ttml-dfxp |  9 +++++----
 tests/ref/fate/mov-mp4-ttml-stpp | 10 ++++++----
 3 files changed, 20 insertions(+), 10 deletions(-)

diff --git a/libavformat/mov.c b/libavformat/mov.c
index 057fd872b1..b8dcd604c1 100644
--- a/libavformat/mov.c
+++ b/libavformat/mov.c
@@ -2264,6 +2264,11 @@ static int mov_codec_id(AVStream *st, uint32_t format)
                     (st->codecpar->codec_type == AVMEDIA_TYPE_SUBTITLE &&
                     st->codecpar->codec_id == AV_CODEC_ID_NONE)) {
             id = ff_codec_get_id(ff_codec_movsubtitle_tags, format);
+            if (id <= 0) {
+                id = (format == MOV_MP4_TTML_TAG || format == MOV_ISMV_TTML_TAG) ?
+                     AV_CODEC_ID_TTML : id;
+            }
+
             if (id > 0)
                 st->codecpar->codec_type = AVMEDIA_TYPE_SUBTITLE;
             else
@@ -2458,8 +2463,10 @@ static void mov_parse_stsd_subtitle(MOVContext *c, AVIOContext *pb,
     // ttxt stsd contains display flags, justification, background
     // color, fonts, and default styles, so fake an atom to read it
     MOVAtom fake_atom = { .size = size };
-    // mp4s contains a regular esds atom
-    if (st->codecpar->codec_tag != AV_RL32("mp4s"))
+    // mp4s contains a regular esds atom, dfxp ISMV TTML has no content
+    // in extradata unlike stpp MP4 TTML.
+    if (st->codecpar->codec_tag != AV_RL32("mp4s") &&
+        st->codecpar->codec_tag != MOV_ISMV_TTML_TAG)
         mov_read_glbl(c, pb, fake_atom);
     st->codecpar->width  = sc->width;
     st->codecpar->height = sc->height;
diff --git a/tests/ref/fate/mov-mp4-ttml-dfxp b/tests/ref/fate/mov-mp4-ttml-dfxp
index 15f4f9b69b..2d60f3d964 100644
--- a/tests/ref/fate/mov-mp4-ttml-dfxp
+++ b/tests/ref/fate/mov-mp4-ttml-dfxp
@@ -1,13 +1,13 @@
 2e7e01c821c111466e7a2844826b7f6d *tests/data/fate/mov-mp4-ttml-dfxp.mp4
 8519 tests/data/fate/mov-mp4-ttml-dfxp.mp4
 #tb 0: 1/1000
-#media_type 0: data
-#codec_id 0: none
+#media_type 0: subtitle
+#codec_id 0: ttml
 0,          0,          0,    68500,     7866, 0x456c36b7
 {
     "packets": [
         {
-            "codec_type": "data",
+            "codec_type": "subtitle",
             "stream_index": 0,
             "pts": 0,
             "pts_time": "0.000000",
@@ -26,7 +26,8 @@
     "streams": [
         {
             "index": 0,
-            "codec_type": "data",
+            "codec_name": "ttml",
+            "codec_type": "subtitle",
             "codec_tag_string": "dfxp",
             "codec_tag": "0x70786664",
             "time_base": "1/1000",
diff --git a/tests/ref/fate/mov-mp4-ttml-stpp b/tests/ref/fate/mov-mp4-ttml-stpp
index 7726cd332e..0a8e423449 100644
--- a/tests/ref/fate/mov-mp4-ttml-stpp
+++ b/tests/ref/fate/mov-mp4-ttml-stpp
@@ -1,13 +1,14 @@
 cbd2c7ff864a663b0d893deac5a0caec *tests/data/fate/mov-mp4-ttml-stpp.mp4
 8547 tests/data/fate/mov-mp4-ttml-stpp.mp4
+#extradata 0:       48, 0x62100c0d
 #tb 0: 1/1000
-#media_type 0: data
-#codec_id 0: none
+#media_type 0: subtitle
+#codec_id 0: ttml
 0,          0,          0,    68500,     7866, 0x456c36b7
 {
     "packets": [
         {
-            "codec_type": "data",
+            "codec_type": "subtitle",
             "stream_index": 0,
             "pts": 0,
             "pts_time": "0.000000",
@@ -26,7 +27,8 @@ cbd2c7ff864a663b0d893deac5a0caec *tests/data/fate/mov-mp4-ttml-stpp.mp4
     "streams": [
         {
             "index": 0,
-            "codec_type": "data",
+            "codec_name": "ttml",
+            "codec_type": "subtitle",
             "codec_tag_string": "stpp",
             "codec_tag": "0x70707473",
             "time_base": "1/1000",
